lib/secretsscanner/scan/scan.go

```
@@ -0,0 +1,252 @@
+		log:         cfg.Log,
+		privateKeys: make(map[string]*accessgraphsecretsv1pb.PrivateKey),
+	}, nil
+}
+
+// Scanner is a scanner that scans directories for secrets.
+type Scanner struct {
+	dirs []string
+	log  *slog.Logger
+	// privateKeys is a map of private keys found during the scan.
+	// The key is the path to the private key file and the value is the private key representation.
+	privateKeys map[string]*accessgraphsecretsv1pb.PrivateKey
+}
+
+// ScanPrivateKeys scans directories for SSH private keys.
+func (s *Scanner) ScanPrivateKeys(ctx context.Context, deviceID string) []SSHPrivateKey {
+	for _, dir := range s.dirs {
+		s.findPrivateKeys(ctx, dir, deviceID)
+	}
+
+	keys := make([]SSHPrivateKey, 0, len(s.privateKeys))
+	for path, key := range s.privateKeys {
+		keys = append(keys, SSHPrivateKey{
+			Path: path,
+			Key:  key,
+		})
+	}
+	return keys
+}
+
+// SSHPrivateKey represents an SSH private key found during the scan.
+type SSHPrivateKey struct {
+	// Path is the absolute path to the private key file.
+	Path string
+	// Key is the private key representation.
+	Key *accessgraphsecretsv1pb.PrivateKey
+}
+
+// findPrivateKeys walks through all files in a directory and its subdirectories
+// and checks if they are SSH private keys.
+func (s *Scanner) findPrivateKeys(ctx context.Context, root, deviceID string) {
+	logger := s.log.With("dir", root)
+
+	err := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			return nil
+		}
+		switch fileData, isKey, err := readFileIfSSHPrivateKey(path); {
+		case err != nil:
+			logger.DebugContext(ctx, "error reading file", "path", path, "error", err)
+		case isKey:
+			key, err := extractSSHKey(ctx, path, deviceID, fileData)
+			if err != nil {
+				logger.DebugContext(ctx, "error extracting private key", "path", path, "error", err)
+			} else {
+				s.privateKeys[path] = key
+			}
+		}
+		return nil
+	})
+
+	if err != nil {
```

As above, don't swallow the error here, let the caller make this choice.